pilot/pkg/model/sidecar.go

```diff
@@ -903,31 +906,67 @@ func (sc *SidecarScope) appendSidecarServices(servicesAdded map[host.Name]sideca
 			return
 		}
 
+		if !canMergeServices(existing, s) {
+			log.Debugf("Service %s/%s from registry %s ignored by %s/%s/%s", s.Attributes.Namespace, s.Hostname, s.Attributes.ServiceRegistry,
+				existing.Attributes.Namespace, existing.Hostname, existing.Attributes.ServiceRegistry)
+			return
+		}
+
 		// we merge ports for services both defined by ServiceEntry in same namespace
-		if existing.Attributes.Namespace == s.Attributes.Namespace {
-			// merge the ports to service when each listener generates partial service
-			// we only merge if the found service is in the same namespace as the one we're trying to add
-			copied := foundSvc.svc.DeepCopy()
-			for _, p := range s.Ports {
-				found := false
-				for _, osp := range copied.Ports {
-					if p.Port == osp.Port {
-						found = true
-						break
-					}
-				}
-				if !found {
-					copied.Ports = append(copied.Ports, p)
+
+		// merge the ports to service when each listener generates partial service
+		// we only merge if the found service is in the same namespace as the one we're trying to add
+		copied := foundSvc.svc.DeepCopy()
+		for _, p := range s.Ports {
+			found := false
+			for _, osp := range copied.Ports {
+				if p.Port == osp.Port {
+					found = true
+					break
 				}
 			}
-			// replace service in slice
-			sc.services[foundSvc.index] = copied
-			// Update index as well, so that future reads will merge into the new service
-			foundSvc.svc = copied
-			servicesAdded[foundSvc.svc.Hostname] = foundSvc
-			sc.servicesByHostname[s.Hostname] = s
+			if !found {
+				copied.Ports = append(copied.Ports, p)
+			}
 		}
+		// replace service in slice
+		sc.services[foundSvc.index] = copied
+		// Update index as well, so that future reads will merge into the new service
+		foundSvc.svc = copied
+		servicesAdded[foundSvc.svc.Hostname] = foundSvc
+		sc.servicesByHostname[s.Hostname] = s
+	}
+}
+
+func canMergeServices(s1, s2 *Service) bool {
+	if s1.Attributes.Namespace != s2.Attributes.Namespace {
+		return false
+	}
+
+	// Hostname has been compared
+
+	if s1.Resolution != s2.Resolution {
+		return false
+	}
+
+	// kuberneres service registry has been checked before
+	if s1.Attributes.ServiceRegistry != s2.Attributes.ServiceRegistry {
+		return false
+	}
+
+	if !maps.Equal(s1.Attributes.Labels, s2.Attributes.Labels) {
+		return false
 	}
+
+	if !maps.Equal(s1.Attributes.LabelSelectors, s2.Attributes.LabelSelectors) {
```

Comment:
hm, do we need this? Seems valid to select different things in different cluster?

Reply:
The original issue https://github.com/istio/istio/issues/49550  has both 
```
  workloadSelector:
    labels:
      app: productpage
```

and 

```
  workloadSelector:
    labels:
      app: ratings
```

They are different services actually, with our current svc merge semantics, we only merge ports, and not the selector. So the result is unexpected

Reply:
Don't we merge the endpoints of them though? The selector is only used to find the endpoints?

Reply:
We cannot get the endpoints for the newer clusters, as the issue reported